app/models/run.rb

```diff
@@ -177,8 +177,10 @@ def previous_pb(timing)
   # instead of individually for each segment (N queries)
   def segment_history_stats(timing)
     stats = SegmentHistory.joins(segment: :run)
+                          .joins(stats_join_additional_histories_query(timing))
                           .where(segment: {runs: {id: id}})
                           .where.not(Run.duration_type(timing) => [0, nil])
+                          .where("segments_segment_histories.segment_number = 0 OR (other_histories.attempt_number = segment_histories.attempt_number AND other_histories.segment_number = segments_segment_histories.segment_number - 1)")
```

Comment:
Which part of the query stops us from getting back segments directly following skipped segments? Are skipped SegmentHistorys absent from the database and the join is enough to filter them out?

Reply:
The `.joins(stats_join_additional_histories_query(timing))` pulls in a second copy of the histories (`other_histories`) that doesn't include histories with a `NULL` or `0` value for the timing we care about so that we can use it for filtering.

`segments_segment_histories` represents the histories we actually care about and use for calculations. `other_histories` (probably not the best name as I sit here and type it out again) are used to be the history immediately prior to `segments_segment_histories`.

The part that stops us from getting back the histories we don't want is the additional `.where`. If the `segment_segment_histories.segment_number` is `0`, then we're the history for the first segment, so we always include that since there's no previous segment to be skipped. Otherwise, we have to look at the segment history of the segment immediately prior to `segments_segment_histories`. We do that with `other_histories.attempt_number = segment_histories.attempt_number AND other_histories.segment_number = segments_segment_histories.segment_number - 1`.

If that history exists in `other_histories`, that means it wasn't skipped, therefore the next history (`segments_segment_histories`) is valid.